### PR TITLE
Added support for coupons

### PR DESCRIPTION
**Pull request by mihailspopovs4**
Added functionality to apply and remove coupon codes.
The result is visible in Cart page and Cart overlay.

This functionality required changes in [scandipwa/quote-graphql](https://github.com/scandipwa/quote-graphql) repository. PR for it will be submitted soon.

### CODE REVIEW

File: src/app/component/CartCoupon/CartCoupon.component.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        );
+    }
+
+    renderRemoveCoupon() {
+        const { couponCode } = this.props;
+
+        return (
+            <>
+                <p block="CartCoupon" elem="Message">
+                    { __('Applied coupon code') }
```
Comment (alfredsgenkins) [inline src/app/component/CartCoupon/CartCoupon.component.js]: `__('Applied coupon code: ')` you can put `:` into translation.

File: src/app/component/CartOverlay/CartOverlay.component.js [hunk 2]
```diff
@@ -56,29 +62,51 @@ export default class CartOverlay extends PureComponent {
...
+        if (!coupon_code) return null;
+
+
+        return (
+            <dl
+              block="CartOverlay"
+              elem="Discount"
+            >
+                <dt>
+                    { __('Coupon') }
```
Comment (alfredsgenkins) [inline src/app/component/CartOverlay/CartOverlay.component.js]: Use: { __('Coupon ') }

File: src/app/component/CartOverlay/CartOverlay.component.js [hunk 3]
```diff
@@ -56,29 +62,51 @@ export default class CartOverlay extends PureComponent {
...
+                    { ' ' }
+                    <strong>{ coupon_code.toUpperCase() }</strong>
+                    { ':' }
```
Comment (alfredsgenkins) [inline src/app/component/CartOverlay/CartOverlay.component.js]: Make this a pseudo-element.

File: src/app/component/CartCoupon/CartCoupon.style.scss [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    &-Message {
+        display: inline-block;
+    }
+
+    &-Input {
+        display: inline-block;
+
+        input {
```
Comment (alfredsgenkins) [inline src/app/component/CartCoupon/CartCoupon.style.scss]: Please check if it is possible to set width to input, without styling the `input` itself.

File: src/app/route/CartPage/CartPage.component.js [hunk 5]
```diff
@@ -112,19 +118,42 @@ class CartPage extends Component {
...
+
+        return (
+            <>
+                <dt>
+                { __('Coupon') }
+                { ' ' }
+                <strong>{ coupon_code.toUpperCase() }</strong>
+                { ':' }
+                </dt>
+                <dd>{ `-${this.renderPriceLine(Math.abs(discount_amount))}` }</dd>
```
Comment (alfredsgenkins) [inline src/app/route/CartPage/CartPage.component.js]: Same as in first comment.

File: src/app/component/CartCoupon/CartCoupon.style.scss (hunk 4 above)
Comment (mihailspopovs4) [reply src/app/component/CartCoupon/CartCoupon.style.scss]: Checked multiple examples where <Field type="text'> is used. Everywhere width is set directly to `input`.

File: src/app/component/CartOverlay/CartOverlay.component.js (hunk 3 above)
Comment (mihailspopovs4) [reply src/app/component/CartOverlay/CartOverlay.component.js]: Resolved

File: src/app/component/CartOverlay/CartOverlay.component.js (hunk 2 above)
Comment (mihailspopovs4) [reply src/app/component/CartOverlay/CartOverlay.component.js]: Resolved

File: src/app/component/CartCoupon/CartCoupon.style.scss (hunk 4 above)
Comment (mihailspopovs4) [reply src/app/component/CartCoupon/CartCoupon.style.scss]: Resolved

File: src/app/route/CartPage/CartPage.component.js (hunk 5 above)
Comment (mihailspopovs4) [reply src/app/route/CartPage/CartPage.component.js]: Resolved

File: src/app/component/CartCoupon/CartCoupon.component.js (hunk 1 above)
Comment (mihailspopovs4) [reply src/app/component/CartCoupon/CartCoupon.component.js]: Resolved

File: src/app/component/CartCoupon/CartCoupon.component.js [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+    state = {
+        enteredCouponCode: ''
+    };
+
+    handleCouponCodeChange = this.handleCouponCodeChange.bind(this);
+
+    handleApplyCoupon = this.handleApplyCoupon.bind(this);
+
+    handleRemoveCoupon = this.handleRemoveCoupon.bind(this);
```
Comment (alfredsgenkins) [inline src/app/component/CartCoupon/CartCoupon.component.js]: Use similar: `renderTableRow = (data) => { ... }`

File: src/app/component/CartCoupon/CartCoupon.component.js [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+    render() {
+        const { isLoading, couponCode } = this.props;
+
+        return (
+            <div block="CartCoupon">
+                <Loader isLoading={ isLoading } />
+                { !couponCode
+                    ? this.renderApplyCoupon()
+                    : this.renderRemoveCoupon() }
```
Comment (alfredsgenkins) [inline src/app/component/CartCoupon/CartCoupon.component.js]: Please move `}` To the next line:

File: src/app/route/CartPage/CartPage.component.js [hunk 8]
```diff
@@ -112,19 +118,41 @@ class CartPage extends Component {
...
+            }
+        } = this.props;
+
+        if (!coupon_code) return null;
+
+        return (
+            <>
+                <dt>
+                    { __('Coupon') }
+                    { ' ' }
```
Comment (alfredsgenkins) [inline src/app/route/CartPage/CartPage.component.js]: `   ` - remove

**Pull request merged**